api/images/services/image_processing_service.py

```diff
@@ -1,16 +1,61 @@
+# from typing import Optional
+
+from io import BytesIO
+
+from django.db import transaction
+
+# from api.users.models import User
+from api.utils.file_storage.image_storage import ImageStorage
 from api.utils.make_temp_directory import make_temp_directory
+from images.models import Analysis, BoundingBox, Image
 
 from .logo_detection_service import LogoDetectionService
 
 
 class ImageProcessingService:
-    def process_images(self, images):
-        detections = {}
+    def __init__(self) -> None:
+        self.storage = ImageStorage()
+
+    def process_images(self, images, user):
+        results = []
 
         with make_temp_directory() as temp_directory:
             for image in images:
-                file_path = temp_directory + "/" + image.name
+                image_file = image.file.read()
+                image_name = image.name
+                file_path = temp_directory + "/" + image_name
                 with open(file_path, "wb") as file:
-                    file.write(image.file.read())
-                detections.update({image.name: LogoDetectionService().detect_in_image(file_path)})
-        return detections
+                    file.write(image_file)
+                detections = LogoDetectionService().detect_in_image(file_path)
+                result = self.__save_to_s3(image_name=image_name, image=image_file, detections=detections, user=user)
+                results.append(result)
+        return results
+
+    # This decorator ensures that the all the create and bulk_create operations are done in one transaction.
+    @transaction.atomic
+    def __save_to_s3(self, image_name, image, detections, user):
+        image_file = BytesIO(image)
+
+        self.storage.save(image_name, image_file)
```

Comment:
Can you put the user's id infront on the image name here, just to avoid collisions of image names:
```suggestion
        self.storage.save(f"{user.id}/{image_name}", image_file)
```